rfcs/20190509-keras-saved-model.md

```diff
@@ -0,0 +1,226 @@
+- `_variables`: List of all variables owned by this object (and not sublayers)
+- `__call__`: Returns the outputs of the call function.
+- `call_and_return_conditional_losses`: Returns the outputs of the call function, as well as a list input-dependent losses (does not include the activity regularizer loss).
+- `activity_regularizer_fn`: Activity regularization function
+- `compile_loss_functions`: List of loss functions added during `model.compile`.
+- `compile_metrics`: List of metric objects added during `model.compile`.
+
+The optimizer is a checkpointable object, so it is automatically saved to the SavedModel.
+
+**Public vs private variables (variables vs _variables)**
+
+The public attributes are exported so that the all variables/trainable variables/etc. may be accessed without the Keras python logic to recursively traverse all the submodules. The private variable attribute is exported so that when the model is deserialized:
+1. it is clear which objects own variables
+2. variables are guaranteed to be in the same order. This is important for Keras models, which uses layer and weight order for certain operations (e.g. saving/loading to HDF5).
+
+**Call functions**
+
+Two versions of the call function are exported. Exporting `__call__` enables the `model(inputs)` function to be retained in the generic object created by `tf.saved_model.load`. `call_and_return_conditional_losses` is exported for model deserialization.
+
+**Common endpoints**
+
+The following properties were proposed by arnoegw@google.com and andresp@google.com, from the TF Hub team, as common endpoints shared by all modules:
+
+- `__call__`: Accepts inputs to the model and returns outputs
+- `variables`: List of all variables in the model
+- `trainable_variables`: List of all trainable variables in the model
+- `regularization_losses`: List of callables that return a scalar tensor.
+
+Having shared endpoints allows models to be used interchangeably between different frameworks. These endpoints are defined above.
+
+### Deserialization details
+The generic loader, `tf.saved_model.load`, creates a generic object with attributes as saved in the SavedModel. Loading the saved Keras model (with added checkpointable objects and functions as listed above) will produce:
+
+```
+GenericObject obj → .variables, .trainable_variables, etc.
+```
+
+`GenericObject` is similar to a Keras model object, but lacks the `.fit`, `.test` and `.predict` methods.
+
+When reconstructing a Keras model, the saved attributes are remapped to the original names. The exception is the call function, which uses `call_and_return_conditional_losses` instead of `__call__`.
+
+## Questions and Discussion Topics
+
+1. Do we all agree on the API changes?
+2. Are there other aspects of the model that should be serialized/deserialized?
+3. Should `tf.saved_model.load` return a generic object or Keras model?
+4. Syncing common endpoints with tf.module
+
```

Comment:
Can you document the backward and forward compatibility guarantees of this serialization format? I assume it'll be fully backwards compatible.

Re forward compatibility, though, what happens if one tries to load a model exported from tf version X into a tf version Y < X? Is the user guaranteed to get all the features that Y supports?

Reply:
Forward compatibility seems feasible. I added a compatibility section promising both backward and forward compatibility.